fwbrasil: first submission

fwbrasil opened the pull request:
#### Check List:
- [x] Tests pass (`./test.sh <username>`)
- [x] All formatting changes by the build are committed
- [x] Output matches that of calculate_average_baseline.sh

* Execution time: 

royvanrijn 0m1.655s
thomaswue 0m1.794s
**fwbrasil 0m1.899s**
obourgain 0m2.219s
ebarlas 0m2.859s
merykitty 0m2.891s

Apple M2 Max

fwbrasil commented:
sorry, I misunderstood the output of the test script. It's not passing, I'm fixing it.

fwbrasil commented:
@gunnarmorling @lobaorn I've fixed all the tests and run the top contenders on my M2 Max. It seems this impl is between the top 3!

gunnarmorling commented on calculate_average_fwbrasil.sh:
```diff
@@ -0,0 +1,23 @@
+#  limitations under the License.
+#
+
+source "$HOME/.sdkman/bin/sdkman-init.sh"
+sdk use java 21.0.1-graal 1>&2
```
Could you move this please to a new script called `prepare_fwbrasil.sh`? We have changed that structure slightly.  The `time` call below also isn't needed any more. Thx.

fwbrasil commented:
I've just updated the scripts. Thank you for working on this! I had a lot of fun and can't wait to try new optimizations this weekend :)

gunnarmorling commented:
Hum, the output of this is different from the baseline. Please run ./calcuate_average_baseline.sh and make sure your output is the same.

fwbrasil commented:
Yes, I had validated that the output matches the baseline. I've just pushed a few optimizations and re-validated it. In case you notice differences again, please share more details. It could something different between mac and linux.

<details>
  <summary>See log</summary>
  fwbrasil@flavios-mbp 1brc % ./test.sh fwbrasil

Using java version 21.0.1-graal in this shell.
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-1.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-10.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-10000-unique-keys.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-2.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-20.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-3.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-boundaries.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-complex-utf8.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-dot.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-short.txt
Validating calculate_average_fwbrasil.sh -- src/test/resources/samples/measurements-shortest.txt

fwbrasil@flavios-mbp 1brc % ./calculate_average_fwbrasil.sh
{Abha=-35.0/18.0/68.2, Abidjan=-22.6/26.0/74.1, Abéché=-20.2/29.4/81.0, Accra=-23.0/26.4/76.1, Addis Ababa=-40.1/16.0/71.2, Adelaide=-31.8/17.3/67.3, Aden=-20.2/29.1/75.8, Ahvaz=-22.5/25.4/78.5, Albuquerque=-32.8/14.0/66.2, Alexandra=-38.8/11.0/62.2, Alexandria=-30.9/20.0/66.4, Algiers=-32.4/18.2/69.5, Alice Springs=-27.4/21.0/72.7, Almaty=-38.0/10.0/59.4, Amsterdam=-39.3/10.2/60.1, Anadyr=-56.2/-6.9/43.6, Anchorage=-46.0/2.8/52.4, Andorra la Vella=-37.6/9.8/58.1, Ankara=-38.9/12.0/58.4, Antananarivo=-33.2/17.9/66.6, Antsiranana=-25.3/25.2/74.2, Arkhangelsk=-48.7/1.3/48.8, Ashgabat=-34.1/17.1/69.4, Asmara=-34.3/15.6/67.3, Assab=-19.5/30.5/79.1, Astana=-50.6/3.5/52.2, Athens=-34.8/19.2/66.6, Atlanta=-31.7/17.0/65.0, Auckland=-36.0/15.2/65.0, Austin=-28.2/20.7/74.6, Baghdad=-28.9/22.8/75.3, Baguio=-29.0/19.5/72.0, Baku=-34.6/15.1/66.2, Baltimore=-35.6/13.1/61.1, Bamako=-23.3/27.8/76.6, Bangkok=-26.8/28.6/78.3, Bangui=-25.0/26.0/79.9, Banjul=-24.2/26.0/72.8, Barcelona=-38.7/18.2/67.1, Bata=-22.8/25.1/78.8, Batumi=-34.0/14.0/61.3, Beijing=-36.1/12.9/59.7, Beirut=-28.9/20.9/71.4, Belgrade=-34.2/12.5/63.3, Belize City=-24.0/26.7/75.7, Benghazi=-29.2/19.9/71.3, Bergen=-40.2/7.7/58.9, Berlin=-38.3/10.3/61.9, Bilbao=-37.5/14.7/68.7, Birao=-23.6/26.5/75.6, Bishkek=-45.2/11.3/60.8, Bissau=-24.1/27.0/76.7, Blantyre=-26.6/22.2/72.1, Bloemfontein=-33.3/15.6/64.9, Boise=-37.6/11.4/61.8, Bordeaux=-31.3/14.2/64.0, Bosaso=-20.9/30.0/79.8, Boston=-38.4/10.9/60.2, Bouaké=-24.1/26.0/75.7, Bratislava=-36.5/10.5/68.1, Brazzaville=-27.9/25.0/74.3, Bridgetown=-22.2/27.0/75.0, Brisbane=-28.9/21.4/71.8, Brussels=-36.5/10.5/60.6, Bucharest=-41.2/10.8/60.7, Budapest=-38.1/11.3/61.6, Bujumbura=-26.3/23.8/77.1, Bulawayo=-29.6/18.9/68.4, Burnie=-37.1/13.1/62.7, Busan=-35.0/15.0/63.5, Cabo San Lucas=-28.7/23.9/76.2, Cairns=-22.2/25.0/73.8, Cairo=-29.7/21.4/75.3, Calgary=-45.1/4.4/52.3, Canberra=-34.4/13.1/62.3, Cape Town=-37.2/16.2/63.5, Changsha=-32.7/17.4/67.5, Charlotte=-33.3/16.1/67.5, Chiang Mai=-21.9/25.8/74.7, Chicago=-42.0/9.8/58.9, Chihuahua=-30.5/18.6/71.1, Chittagong=-25.5/25.9/75.6, Chișinău=-38.7/10.2/58.9, Chongqing=-35.4/18.6/70.5, Christchurch=-37.0/12.2/63.1, City of San Marino=-36.8/11.8/65.1, Colombo=-29.5/27.4/75.3, Columbus=-38.3/11.7/61.6, Conakry=-25.5/26.4/80.2, Copenhagen=-39.8/9.1/58.6, Cotonou=-20.4/27.2/75.2, Cracow=-41.7/9.3/64.3, Da Lat=-33.5/17.9/68.9, Da Nang=-25.3/25.8/73.3, Dakar=-28.3/24.0/75.3, Dallas=-30.6/19.0/67.8, Damascus=-30.6/17.0/71.8, Dampier=-24.4/26.4/76.3, Dar es Salaam=-22.1/25.8/74.3, Darwin=-19.9/27.6/81.1, Denpasar=-25.8/23.7/74.6, Denver=-41.7/10.4/61.2, Detroit=-38.4/10.0/59.0, Dhaka=-24.9/25.9/78.1, Dikson=-66.5/-11.1/39.4, Dili=-25.2/26.6/76.7, Djibouti=-20.0/29.9/82.2, Dodoma=-26.3/22.7/74.9, Dolisie=-23.6/24.0/76.1, Douala=-22.2/26.7/76.6, Dubai=-21.5/26.9/76.4, Dublin=-40.6/9.8/63.0, Dunedin=-37.7/11.1/65.0, Durban=-30.1/20.6/75.6, Dushanbe=-37.5/14.7/64.2, Edinburgh=-40.1/9.3/60.4, Edmonton=-44.8/4.2/60.3, El Paso=-32.4/18.1/68.9, Entebbe=-28.1/21.0/71.2, Erbil=-29.7/19.5/68.0, Erzurum=-47.6/5.1/58.7, Fairbanks=-51.2/-2.3/47.4, Fianarantsoa=-32.9/17.9/69.1, Flores,  Petén=-21.2/26.4/81.8, Frankfurt=-44.6/10.6/64.7, Fresno=-33.2/17.9/68.5, Fukuoka=-31.6/17.0/71.4, Gaborone=-27.8/21.0/70.0, Gabès=-28.1/19.5/69.8, Gagnoa=-23.8/26.0/78.6, Gangtok=-37.4/15.2/62.7, Garissa=-20.4/29.3/81.1, Garoua=-20.7/28.3/78.0, George Town=-19.6/27.9/77.7, Ghanzi=-30.7/21.4/75.6, Gjoa Haven=-65.1/-14.4/42.9, Guadalajara=-31.1/20.9/70.6, Guangzhou=-24.8/22.4/76.0, Guatemala City=-27.5/20.4/69.3, Halifax=-46.2/7.5/60.8, Hamburg=-36.2/9.7/60.9, Hamilton=-36.9/13.8/62.1, Hanga Roa=-28.4/20.5/73.8, Hanoi=-27.1/23.6/73.6, Harare=-33.3/18.4/67.7, Harbin=-43.8/5.0/55.4, Hargeisa=-27.0/21.7/68.6, Hat Yai=-23.9/27.0/76.8, Havana=-30.6/25.2/73.6, Helsinki=-42.0/5.9/57.8, Heraklion=-34.7/18.9/70.5, Hiroshima=-30.7/16.3/68.2, Ho Chi Minh City=-19.4/27.4/76.7, Hobart=-41.2/12.7/60.6, Hong Kong=-29.2/23.3/72.5, Honiara=-22.6/26.5/75.3, Honolulu=-24.7/25.4/79.7, Houston=-27.2/20.8/71.0, Ifrane=-45.0/11.4/60.4, Indianapolis=-37.0/11.8/60.0, Iqaluit=-62.2/-9.3/39.6, Irkutsk=-52.7/1.0/53.0, Istanbul=-36.5/13.9/63.4, Jacksonville=-25.5/20.3/72.2, Jakarta=-26.3/26.7/77.6, Jayapura=-23.3/27.0/76.1, Jerusalem=-30.8/18.3/67.3, Johannesburg=-32.1/15.5/64.8, Jos=-27.9/22.8/71.2, Juba=-20.6/27.8/77.3, Kabul=-38.0/12.1/64.3, Kampala=-29.9/20.0/70.1, Kandi=-25.2/27.7/76.1, Kankan=-25.6/26.5/78.1, Kano=-23.9/26.4/75.9, Kansas City=-39.0/12.5/63.8, Karachi=-34.3/26.0/74.8, Karonga=-26.7/24.4/76.7, Kathmandu=-29.6/18.3/68.4, Khartoum=-18.3/29.9/78.0, Kingston=-18.7/27.4/81.3, Kinshasa=-23.9/25.3/77.4, Kolkata=-22.7/26.7/75.1, Kuala Lumpur=-25.0/27.3/79.1, Kumasi=-23.1/26.0/77.0, Kunming=-34.3/15.7/65.7, Kuopio=-47.2/3.4/52.0, Kuwait City=-23.6/25.7/79.2, Kyiv=-40.1/8.4/58.1, Kyoto=-31.6/15.8/62.8, La Ceiba=-23.7/26.2/76.3, La Paz=-31.3/23.7/73.2, Lagos=-27.8/26.8/87.8, Lahore=-28.8/24.3/74.1, Lake Havasu City=-25.4/23.7/77.6, Lake Tekapo=-37.5/8.7/61.7, Las Palmas de Gran Canaria=-24.9/21.2/70.3, Las Vegas=-26.5/20.3/69.8, Launceston=-42.9/13.1/64.0, Lhasa=-43.2/7.6/58.1, Libreville=-24.1/25.9/73.1, Lisbon=-32.2/17.5/69.1, Livingstone=-30.4/21.8/71.1, Ljubljana=-38.1/10.9/59.3, Lodwar=-19.0/29.3/76.5, Lomé=-29.2/26.9/80.4, London=-36.6/11.3/59.1, Los Angeles=-33.8/18.6/69.0, Louisville=-35.6/13.9/64.9, Luanda=-23.8/25.8/77.3, Lubumbashi=-27.6/20.8/72.0, Lusaka=-37.3/19.9/72.8, Luxembourg City=-38.9/9.3/57.5, Lviv=-41.2/7.8/58.4, Lyon=-38.2/12.5/60.3, Madrid=-35.0/15.0/66.2, Mahajanga=-26.1/26.3/78.6, Makassar=-23.2/26.7/76.6, Makurdi=-22.8/26.0/73.8, Malabo=-23.0/26.3/77.1, Malé=-24.7/28.0/78.9, Managua=-18.9/27.3/77.0, Manama=-23.1/26.5/78.6, Mandalay=-21.2/28.0/76.1, Mango=-19.7/28.1/78.5, Manila=-19.0/28.4/81.2, Maputo=-35.9/22.8/73.1, Marrakesh=-29.6/19.6/68.8, Marseille=-34.5/15.8/66.0, Maun=-24.8/22.4/76.4, Medan=-23.2/26.5/79.7, Mek'ele=-31.8/22.7/69.2, Melbourne=-35.3/15.1/61.2, Memphis=-36.2/17.2/68.1, Mexicali=-26.0/23.1/71.6, Mexico City=-33.8/17.5/66.1, Miami=-23.3/24.9/73.1, Milan=-38.7/13.0/62.6, Milwaukee=-40.0/8.9/56.6, Minneapolis=-39.3/7.8/61.0, Minsk=-43.4/6.7/53.8, Mogadishu=-26.5/27.1/83.5, Mombasa=-28.3/26.3/73.5, Monaco=-38.1/16.4/64.4, Moncton=-47.7/6.1/60.4, Monterrey=-25.9/22.3/71.8, Montreal=-48.2/6.8/57.3, Moscow=-49.5/5.8/55.7, Mumbai=-22.5/27.1/79.6, Murmansk=-47.8/0.6/47.8, Muscat=-25.7/28.0/77.2, Mzuzu=-34.7/17.7/70.3, N'Djamena=-19.5/28.3/77.3, Naha=-29.7/23.1/78.3, Nairobi=-31.0/17.8/69.3, Nakhon Ratchasima=-24.4/27.3/76.6, Napier=-36.3/14.6/62.1, Napoli=-39.8/15.9/69.9, Nashville=-44.5/15.4/65.6, Nassau=-27.5/24.6/79.1, Ndola=-29.7/20.3/70.4, New Delhi=-26.5/25.0/74.9, New Orleans=-30.5/20.7/68.9, New York City=-39.5/12.9/59.2, Ngaoundéré=-30.5/22.0/72.0, Niamey=-17.7/29.3/77.6, Nicosia=-28.9/19.7/66.7, Niigata=-38.6/13.9/67.2, Nouadhibou=-28.1/21.3/70.3, Nouakchott=-25.2/25.7/76.0, Novosibirsk=-44.7/1.7/50.2, Nuuk=-48.8/-1.4/56.2, Odesa=-38.3/10.7/63.3, Odienné=-22.7/26.0/75.0, Oklahoma City=-31.5/15.9/64.6, Omaha=-40.5/10.6/61.6, Oranjestad=-20.7/28.1/76.4, Oslo=-43.1/5.7/56.4, Ottawa=-48.3/6.6/56.3, Ouagadougou=-22.3/28.3/76.5, Ouahigouya=-21.7/28.6/77.6, Ouarzazate=-30.6/18.9/68.0, Oulu=-46.1/2.7/53.7, Palembang=-23.6/27.3/75.5, Palermo=-32.8/18.5/65.6, Palm Springs=-23.8/24.5/77.3, Palmerston North=-33.6/13.2/64.5, Panama City=-22.2/28.0/79.0, Parakou=-22.6/26.8/77.9, Paris=-37.5/12.3/59.0, Perth=-29.5/18.7/67.2, Petropavlovsk-Kamchatsky=-46.9/1.9/51.3, Philadelphia=-33.4/13.2/63.8, Phnom Penh=-22.9/28.3/77.6, Phoenix=-24.7/23.9/75.7, Pittsburgh=-36.9/10.8/61.8, Podgorica=-33.7/15.3/63.1, Pointe-Noire=-27.6/26.1/78.3, Pontianak=-25.7/27.7/75.4, Port Moresby=-20.3/26.9/77.0, Port Sudan=-24.2/28.4/78.4, Port Vila=-26.0/24.3/74.9, Port-Gentil=-24.5/26.0/77.8, Portland (OR)=-38.2/12.4/61.8, Porto=-39.7/15.7/65.9, Prague=-47.7/8.4/61.7, Praia=-25.7/24.4/75.6, Pretoria=-30.9/18.2/68.3, Pyongyang=-39.3/10.8/68.2, Rabat=-32.6/17.2/71.2, Rangpur=-27.5/24.4/74.2, Reggane=-20.1/28.3/80.0, Reykjavík=-47.4/4.3/55.4, Riga=-41.7/6.2/57.9, Riyadh=-24.0/26.0/74.2, Rome=-32.4/15.2/66.4, Roseau=-21.7/26.2/72.7, Rostov-on-Don=-39.6/9.9/59.7, Sacramento=-33.3/16.3/66.0, Saint Petersburg=-45.1/5.8/55.2, Saint-Pierre=-46.7/5.7/60.8, Salt Lake City=-40.3/11.6/64.4, San Antonio=-26.8/20.8/73.0, San Diego=-30.0/17.8/69.2, San Francisco=-35.4/14.6/72.2, San Jose=-38.9/16.4/69.0, San José=-29.0/22.6/73.2, San Juan=-22.0/27.2/78.5, San Salvador=-26.7/23.1/80.2, Sana'a=-31.9/20.0/70.1, Santo Domingo=-23.4/25.9/78.9, Sapporo=-41.7/8.9/59.7, Sarajevo=-39.6/10.1/61.6, Saskatoon=-46.6/3.3/51.2, Seattle=-40.2/11.3/60.8, Seoul=-38.3/12.5/59.4, Seville=-31.9/19.2/68.2, Shanghai=-31.7/16.7/69.1, Singapore=-21.8/27.0/74.7, Skopje=-39.8/12.4/61.4, Sochi=-32.5/14.2/69.0, Sofia=-38.7/10.6/61.3, Sokoto=-23.8/28.0/78.0, Split=-31.3/16.1/65.2, St. John's=-44.9/5.0/53.8, St. Louis=-37.5/13.9/64.1, Stockholm=-42.9/6.6/59.3, Surabaya=-25.3/27.1/76.9, Suva=-25.6/25.6/74.1, Suwałki=-43.3/7.2/55.1, Sydney=-32.7/17.7/67.5, Ségou=-19.0/28.0/81.1, Tabora=-28.2/23.0/69.0, Tabriz=-35.5/12.6/65.2, Taipei=-29.1/23.0/73.8, Tallinn=-42.7/6.4/56.7, Tamale=-23.2/27.9/77.3, Tamanrasset=-28.0/21.7/72.7, Tampa=-29.3/22.9/77.2, Tashkent=-34.0/14.8/64.6, Tauranga=-36.3/14.8/64.8, Tbilisi=-36.0/12.9/65.4, Tegucigalpa=-28.6/21.7/74.7, Tehran=-33.9/17.0/67.2, Tel Aviv=-28.4/20.0/68.8, Thessaloniki=-30.6/16.0/72.3, Thiès=-25.8/24.0/76.8, Tijuana=-31.2/17.8/70.5, Timbuktu=-21.6/28.0/78.1, Tirana=-39.4/15.2/65.9, Toamasina=-28.4/23.4/72.9, Tokyo=-35.9/15.4/66.3, Toliara=-23.8/24.1/72.7, Toluca=-37.6/12.4/65.8, Toronto=-39.2/9.4/57.9, Tripoli=-28.9/20.0/71.0, Tromsø=-49.0/2.9/52.9, Tucson=-28.1/20.9/73.9, Tunis=-32.3/18.4/65.3, Ulaanbaatar=-49.4/-0.4/48.1, Upington=-27.3/20.4/71.1, Vaduz=-39.7/10.1/57.9, Valencia=-37.1/18.3/69.6, Valletta=-32.5/18.8/67.6, Vancouver=-38.3/10.4/61.0, Veracruz=-24.8/25.4/74.8, Vienna=-37.2/10.4/66.5, Vientiane=-30.6/25.9/76.7, Villahermosa=-21.4/27.1/77.5, Vilnius=-44.6/6.0/54.9, Virginia Beach=-34.4/15.8/66.8, Vladivostok=-44.8/4.9/53.9, Warsaw=-44.1/8.5/59.1, Washington, D.C.=-36.0/14.6/67.7, Wau=-22.8/27.8/76.8, Wellington=-38.5/12.9/61.9, Whitehorse=-50.6/-0.1/50.0, Wichita=-36.5/13.9/60.1, Willemstad=-22.3/28.0/76.5, Winnipeg=-46.6/3.0/55.8, Wrocław=-41.1/9.6/64.7, Xi'an=-36.6/14.1/63.7, Yakutsk=-61.3/-8.8/41.2, Yangon=-25.2/27.5/78.5, Yaoundé=-25.1/23.8/79.5, Yellowknife=-56.5/-4.3/44.3, Yerevan=-39.7/12.4/62.1, Yinchuan=-37.9/9.0/61.8, Zagreb=-38.6/10.7/60.9, Zanzibar City=-24.9/26.0/76.6, Zürich=-41.5/9.3/55.6, Ürümqi=-41.8/7.4/60.4, İzmir=-32.1/17.9/68.0}

fwbrasil@flavios-mbp 1brc % ./calculate_average_baseline.sh
{Abha=-35.0/18.0/68.2, Abidjan=-22.6/26.0/74.1, Abéché=-20.2/29.4/81.0, Accra=-23.0/26.4/76.1, Addis Ababa=-40.1/16.0/71.2, Adelaide=-31.8/17.3/67.3, Aden=-20.2/29.1/75.8, Ahvaz=-22.5/25.4/78.5, Albuquerque=-32.8/14.0/66.2, Alexandra=-38.8/11.0/62.2, Alexandria=-30.9/20.0/66.4, Algiers=-32.4/18.2/69.5, Alice Springs=-27.4/21.0/72.7, Almaty=-38.0/10.0/59.4, Amsterdam=-39.3/10.2/60.1, Anadyr=-56.2/-6.9/43.6, Anchorage=-46.0/2.8/52.4, Andorra la Vella=-37.6/9.8/58.1, Ankara=-38.9/12.0/58.4, Antananarivo=-33.2/17.9/66.6, Antsiranana=-25.3/25.2/74.2, Arkhangelsk=-48.7/1.3/48.8, Ashgabat=-34.1/17.1/69.4, Asmara=-34.3/15.6/67.3, Assab=-19.5/30.5/79.1, Astana=-50.6/3.5/52.2, Athens=-34.8/19.2/66.6, Atlanta=-31.7/17.0/65.0, Auckland=-36.0/15.2/65.0, Austin=-28.2/20.7/74.6, Baghdad=-28.9/22.8/75.3, Baguio=-29.0/19.5/72.0, Baku=-34.6/15.1/66.2, Baltimore=-35.6/13.1/61.1, Bamako=-23.3/27.8/76.6, Bangkok=-26.8/28.6/78.3, Bangui=-25.0/26.0/79.9, Banjul=-24.2/26.0/72.8, Barcelona=-38.7/18.2/67.1, Bata=-22.8/25.1/78.8, Batumi=-34.0/14.0/61.3, Beijing=-36.1/12.9/59.7, Beirut=-28.9/20.9/71.4, Belgrade=-34.2/12.5/63.3, Belize City=-24.0/26.7/75.7, Benghazi=-29.2/19.9/71.3, Bergen=-40.2/7.7/58.9, Berlin=-38.3/10.3/61.9, Bilbao=-37.5/14.7/68.7, Birao=-23.6/26.5/75.6, Bishkek=-45.2/11.3/60.8, Bissau=-24.1/27.0/76.7, Blantyre=-26.6/22.2/72.1, Bloemfontein=-33.3/15.6/64.9, Boise=-37.6/11.4/61.8, Bordeaux=-31.3/14.2/64.0, Bosaso=-20.9/30.0/79.8, Boston=-38.4/10.9/60.2, Bouaké=-24.1/26.0/75.7, Bratislava=-36.5/10.5/68.1, Brazzaville=-27.9/25.0/74.3, Bridgetown=-22.2/27.0/75.0, Brisbane=-28.9/21.4/71.8, Brussels=-36.5/10.5/60.6, Bucharest=-41.2/10.8/60.7, Budapest=-38.1/11.3/61.6, Bujumbura=-26.3/23.8/77.1, Bulawayo=-29.6/18.9/68.4, Burnie=-37.1/13.1/62.7, Busan=-35.0/15.0/63.5, Cabo San Lucas=-28.7/23.9/76.2, Cairns=-22.2/25.0/73.8, Cairo=-29.7/21.4/75.3, Calgary=-45.1/4.4/52.3, Canberra=-34.4/13.1/62.3, Cape Town=-37.2/16.2/63.5, Changsha=-32.7/17.4/67.5, Charlotte=-33.3/16.1/67.5, Chiang Mai=-21.9/25.8/74.7, Chicago=-42.0/9.8/58.9, Chihuahua=-30.5/18.6/71.1, Chittagong=-25.5/25.9/75.6, Chișinău=-38.7/10.2/58.9, Chongqing=-35.4/18.6/70.5, Christchurch=-37.0/12.2/63.1, City of San Marino=-36.8/11.8/65.1, Colombo=-29.5/27.4/75.3, Columbus=-38.3/11.7/61.6, Conakry=-25.5/26.4/80.2, Copenhagen=-39.8/9.1/58.6, Cotonou=-20.4/27.2/75.2, Cracow=-41.7/9.3/64.3, Da Lat=-33.5/17.9/68.9, Da Nang=-25.3/25.8/73.3, Dakar=-28.3/24.0/75.3, Dallas=-30.6/19.0/67.8, Damascus=-30.6/17.0/71.8, Dampier=-24.4/26.4/76.3, Dar es Salaam=-22.1/25.8/74.3, Darwin=-19.9/27.6/81.1, Denpasar=-25.8/23.7/74.6, Denver=-41.7/10.4/61.2, Detroit=-38.4/10.0/59.0, Dhaka=-24.9/25.9/78.1, Dikson=-66.5/-11.1/39.4, Dili=-25.2/26.6/76.7, Djibouti=-20.0/29.9/82.2, Dodoma=-26.3/22.7/74.9, Dolisie=-23.6/24.0/76.1, Douala=-22.2/26.7/76.6, Dubai=-21.5/26.9/76.4, Dublin=-40.6/9.8/63.0, Dunedin=-37.7/11.1/65.0, Durban=-30.1/20.6/75.6, Dushanbe=-37.5/14.7/64.2, Edinburgh=-40.1/9.3/60.4, Edmonton=-44.8/4.2/60.3, El Paso=-32.4/18.1/68.9, Entebbe=-28.1/21.0/71.2, Erbil=-29.7/19.5/68.0, Erzurum=-47.6/5.1/58.7, Fairbanks=-51.2/-2.3/47.4, Fianarantsoa=-32.9/17.9/69.1, Flores,  Petén=-21.2/26.4/81.8, Frankfurt=-44.6/10.6/64.7, Fresno=-33.2/17.9/68.5, Fukuoka=-31.6/17.0/71.4, Gaborone=-27.8/21.0/70.0, Gabès=-28.1/19.5/69.8, Gagnoa=-23.8/26.0/78.6, Gangtok=-37.4/15.2/62.7, Garissa=-20.4/29.3/81.1, Garoua=-20.7/28.3/78.0, George Town=-19.6/27.9/77.7, Ghanzi=-30.7/21.4/75.6, Gjoa Haven=-65.1/-14.4/42.9, Guadalajara=-31.1/20.9/70.6, Guangzhou=-24.8/22.4/76.0, Guatemala City=-27.5/20.4/69.3, Halifax=-46.2/7.5/60.8, Hamburg=-36.2/9.7/60.9, Hamilton=-36.9/13.8/62.1, Hanga Roa=-28.4/20.5/73.8, Hanoi=-27.1/23.6/73.6, Harare=-33.3/18.4/67.7, Harbin=-43.8/5.0/55.4, Hargeisa=-27.0/21.7/68.6, Hat Yai=-23.9/27.0/76.8, Havana=-30.6/25.2/73.6, Helsinki=-42.0/5.9/57.8, Heraklion=-34.7/18.9/70.5, Hiroshima=-30.7/16.3/68.2, Ho Chi Minh City=-19.4/27.4/76.7, Hobart=-41.2/12.7/60.6, Hong Kong=-29.2/23.3/72.5, Honiara=-22.6/26.5/75.3, Honolulu=-24.7/25.4/79.7, Houston=-27.2/20.8/71.0, Ifrane=-45.0/11.4/60.4, Indianapolis=-37.0/11.8/60.0, Iqaluit=-62.2/-9.3/39.6, Irkutsk=-52.7/1.0/53.0, Istanbul=-36.5/13.9/63.4, Jacksonville=-25.5/20.3/72.2, Jakarta=-26.3/26.7/77.6, Jayapura=-23.3/27.0/76.1, Jerusalem=-30.8/18.3/67.3, Johannesburg=-32.1/15.5/64.8, Jos=-27.9/22.8/71.2, Juba=-20.6/27.8/77.3, Kabul=-38.0/12.1/64.3, Kampala=-29.9/20.0/70.1, Kandi=-25.2/27.7/76.1, Kankan=-25.6/26.5/78.1, Kano=-23.9/26.4/75.9, Kansas City=-39.0/12.5/63.8, Karachi=-34.3/26.0/74.8, Karonga=-26.7/24.4/76.7, Kathmandu=-29.6/18.3/68.4, Khartoum=-18.3/29.9/78.0, Kingston=-18.7/27.4/81.3, Kinshasa=-23.9/25.3/77.4, Kolkata=-22.7/26.7/75.1, Kuala Lumpur=-25.0/27.3/79.1, Kumasi=-23.1/26.0/77.0, Kunming=-34.3/15.7/65.7, Kuopio=-47.2/3.4/52.0, Kuwait City=-23.6/25.7/79.2, Kyiv=-40.1/8.4/58.1, Kyoto=-31.6/15.8/62.8, La Ceiba=-23.7/26.2/76.3, La Paz=-31.3/23.7/73.2, Lagos=-27.8/26.8/87.8, Lahore=-28.8/24.3/74.1, Lake Havasu City=-25.4/23.7/77.6, Lake Tekapo=-37.5/8.7/61.7, Las Palmas de Gran Canaria=-24.9/21.2/70.3, Las Vegas=-26.5/20.3/69.8, Launceston=-42.9/13.1/64.0, Lhasa=-43.2/7.6/58.1, Libreville=-24.1/25.9/73.1, Lisbon=-32.2/17.5/69.1, Livingstone=-30.4/21.8/71.1, Ljubljana=-38.1/10.9/59.3, Lodwar=-19.0/29.3/76.5, Lomé=-29.2/26.9/80.4, London=-36.6/11.3/59.1, Los Angeles=-33.8/18.6/69.0, Louisville=-35.6/13.9/64.9, Luanda=-23.8/25.8/77.3, Lubumbashi=-27.6/20.8/72.0, Lusaka=-37.3/19.9/72.8, Luxembourg City=-38.9/9.3/57.5, Lviv=-41.2/7.8/58.4, Lyon=-38.2/12.5/60.3, Madrid=-35.0/15.0/66.2, Mahajanga=-26.1/26.3/78.6, Makassar=-23.2/26.7/76.6, Makurdi=-22.8/26.0/73.8, Malabo=-23.0/26.3/77.1, Malé=-24.7/28.0/78.9, Managua=-18.9/27.3/77.0, Manama=-23.1/26.5/78.6, Mandalay=-21.2/28.0/76.1, Mango=-19.7/28.1/78.5, Manila=-19.0/28.4/81.2, Maputo=-35.9/22.8/73.1, Marrakesh=-29.6/19.6/68.8, Marseille=-34.5/15.8/66.0, Maun=-24.8/22.4/76.4, Medan=-23.2/26.5/79.7, Mek'ele=-31.8/22.7/69.2, Melbourne=-35.3/15.1/61.2, Memphis=-36.2/17.2/68.1, Mexicali=-26.0/23.1/71.6, Mexico City=-33.8/17.5/66.1, Miami=-23.3/24.9/73.1, Milan=-38.7/13.0/62.6, Milwaukee=-40.0/8.9/56.6, Minneapolis=-39.3/7.8/61.0, Minsk=-43.4/6.7/53.8, Mogadishu=-26.5/27.1/83.5, Mombasa=-28.3/26.3/73.5, Monaco=-38.1/16.4/64.4, Moncton=-47.7/6.1/60.4, Monterrey=-25.9/22.3/71.8, Montreal=-48.2/6.8/57.3, Moscow=-49.5/5.8/55.7, Mumbai=-22.5/27.1/79.6, Murmansk=-47.8/0.6/47.8, Muscat=-25.7/28.0/77.2, Mzuzu=-34.7/17.7/70.3, N'Djamena=-19.5/28.3/77.3, Naha=-29.7/23.1/78.3, Nairobi=-31.0/17.8/69.3, Nakhon Ratchasima=-24.4/27.3/76.6, Napier=-36.3/14.6/62.1, Napoli=-39.8/15.9/69.9, Nashville=-44.5/15.4/65.6, Nassau=-27.5/24.6/79.1, Ndola=-29.7/20.3/70.4, New Delhi=-26.5/25.0/74.9, New Orleans=-30.5/20.7/68.9, New York City=-39.5/12.9/59.2, Ngaoundéré=-30.5/22.0/72.0, Niamey=-17.7/29.3/77.6, Nicosia=-28.9/19.7/66.7, Niigata=-38.6/13.9/67.2, Nouadhibou=-28.1/21.3/70.3, Nouakchott=-25.2/25.7/76.0, Novosibirsk=-44.7/1.7/50.2, Nuuk=-48.8/-1.4/56.2, Odesa=-38.3/10.7/63.3, Odienné=-22.7/26.0/75.0, Oklahoma City=-31.5/15.9/64.6, Omaha=-40.5/10.6/61.6, Oranjestad=-20.7/28.1/76.4, Oslo=-43.1/5.7/56.4, Ottawa=-48.3/6.6/56.3, Ouagadougou=-22.3/28.3/76.5, Ouahigouya=-21.7/28.6/77.6, Ouarzazate=-30.6/18.9/68.0, Oulu=-46.1/2.7/53.7, Palembang=-23.6/27.3/75.5, Palermo=-32.8/18.5/65.6, Palm Springs=-23.8/24.5/77.3, Palmerston North=-33.6/13.2/64.5, Panama City=-22.2/28.0/79.0, Parakou=-22.6/26.8/77.9, Paris=-37.5/12.3/59.0, Perth=-29.5/18.7/67.2, Petropavlovsk-Kamchatsky=-46.9/1.9/51.3, Philadelphia=-33.4/13.2/63.8, Phnom Penh=-22.9/28.3/77.6, Phoenix=-24.7/23.9/75.7, Pittsburgh=-36.9/10.8/61.8, Podgorica=-33.7/15.3/63.1, Pointe-Noire=-27.6/26.1/78.3, Pontianak=-25.7/27.7/75.4, Port Moresby=-20.3/26.9/77.0, Port Sudan=-24.2/28.4/78.4, Port Vila=-26.0/24.3/74.9, Port-Gentil=-24.5/26.0/77.8, Portland (OR)=-38.2/12.4/61.8, Porto=-39.7/15.7/65.9, Prague=-47.7/8.4/61.7, Praia=-25.7/24.4/75.6, Pretoria=-30.9/18.2/68.3, Pyongyang=-39.3/10.8/68.2, Rabat=-32.6/17.2/71.2, Rangpur=-27.5/24.4/74.2, Reggane=-20.1/28.3/80.0, Reykjavík=-47.4/4.3/55.4, Riga=-41.7/6.2/57.9, Riyadh=-24.0/26.0/74.2, Rome=-32.4/15.2/66.4, Roseau=-21.7/26.2/72.7, Rostov-on-Don=-39.6/9.9/59.7, Sacramento=-33.3/16.3/66.0, Saint Petersburg=-45.1/5.8/55.2, Saint-Pierre=-46.7/5.7/60.8, Salt Lake City=-40.3/11.6/64.4, San Antonio=-26.8/20.8/73.0, San Diego=-30.0/17.8/69.2, San Francisco=-35.4/14.6/72.2, San Jose=-38.9/16.4/69.0, San José=-29.0/22.6/73.2, San Juan=-22.0/27.2/78.5, San Salvador=-26.7/23.1/80.2, Sana'a=-31.9/20.0/70.1, Santo Domingo=-23.4/25.9/78.9, Sapporo=-41.7/8.9/59.7, Sarajevo=-39.6/10.1/61.6, Saskatoon=-46.6/3.3/51.2, Seattle=-40.2/11.3/60.8, Seoul=-38.3/12.5/59.4, Seville=-31.9/19.2/68.2, Shanghai=-31.7/16.7/69.1, Singapore=-21.8/27.0/74.7, Skopje=-39.8/12.4/61.4, Sochi=-32.5/14.2/69.0, Sofia=-38.7/10.6/61.3, Sokoto=-23.8/28.0/78.0, Split=-31.3/16.1/65.2, St. John's=-44.9/5.0/53.8, St. Louis=-37.5/13.9/64.1, Stockholm=-42.9/6.6/59.3, Surabaya=-25.3/27.1/76.9, Suva=-25.6/25.6/74.1, Suwałki=-43.3/7.2/55.1, Sydney=-32.7/17.7/67.5, Ségou=-19.0/28.0/81.1, Tabora=-28.2/23.0/69.0, Tabriz=-35.5/12.6/65.2, Taipei=-29.1/23.0/73.8, Tallinn=-42.7/6.4/56.7, Tamale=-23.2/27.9/77.3, Tamanrasset=-28.0/21.7/72.7, Tampa=-29.3/22.9/77.2, Tashkent=-34.0/14.8/64.6, Tauranga=-36.3/14.8/64.8, Tbilisi=-36.0/12.9/65.4, Tegucigalpa=-28.6/21.7/74.7, Tehran=-33.9/17.0/67.2, Tel Aviv=-28.4/20.0/68.8, Thessaloniki=-30.6/16.0/72.3, Thiès=-25.8/24.0/76.8, Tijuana=-31.2/17.8/70.5, Timbuktu=-21.6/28.0/78.1, Tirana=-39.4/15.2/65.9, Toamasina=-28.4/23.4/72.9, Tokyo=-35.9/15.4/66.3, Toliara=-23.8/24.1/72.7, Toluca=-37.6/12.4/65.8, Toronto=-39.2/9.4/57.9, Tripoli=-28.9/20.0/71.0, Tromsø=-49.0/2.9/52.9, Tucson=-28.1/20.9/73.9, Tunis=-32.3/18.4/65.3, Ulaanbaatar=-49.4/-0.4/48.1, Upington=-27.3/20.4/71.1, Vaduz=-39.7/10.1/57.9, Valencia=-37.1/18.3/69.6, Valletta=-32.5/18.8/67.6, Vancouver=-38.3/10.4/61.0, Veracruz=-24.8/25.4/74.8, Vienna=-37.2/10.4/66.5, Vientiane=-30.6/25.9/76.7, Villahermosa=-21.4/27.1/77.5, Vilnius=-44.6/6.0/54.9, Virginia Beach=-34.4/15.8/66.8, Vladivostok=-44.8/4.9/53.9, Warsaw=-44.1/8.5/59.1, Washington, D.C.=-36.0/14.6/67.7, Wau=-22.8/27.8/76.8, Wellington=-38.5/12.9/61.9, Whitehorse=-50.6/-0.1/50.0, Wichita=-36.5/13.9/60.1, Willemstad=-22.3/28.0/76.5, Winnipeg=-46.6/3.0/55.8, Wrocław=-41.1/9.6/64.7, Xi'an=-36.6/14.1/63.7, Yakutsk=-61.3/-8.8/41.2, Yangon=-25.2/27.5/78.5, Yaoundé=-25.1/23.8/79.5, Yellowknife=-56.5/-4.3/44.3, Yerevan=-39.7/12.4/62.1, Yinchuan=-37.9/9.0/61.8, Zagreb=-38.6/10.7/60.9, Zanzibar City=-24.9/26.0/76.6, Zürich=-41.5/9.3/55.6, Ürümqi=-41.8/7.4/60.4, İzmir=-32.1/17.9/68.0}

</details>

gunnarmorling commented on src/main/java/dev/morling/onebrc/CalculateAverage_fwbrasil.java:
```diff
@@ -0,0 +1,383 @@
+            }
+        }
+
+        int findKey(long keyStart, long keyEnd, long hash) {
```
Where do you deal with hash collisions?

fwbrasil replied:
The goal was to find a good hash function without conflicts in the long range for the workload and tests. Is that not allowed? I'm getting access to a linux box today to work on the benchmark in case you're still seeing differences.

gunnarmorling replied:
As per the rules, it must work for any station names of the allowed length:

> Q: Can I make assumptions on the names of the weather stations showing up in the data set?
A: No, while only a fixed set of station names is used by the data set generator, any solution should work with arbitrary UTF-8 station names (for the sake of simplicity, names are guaranteed to contain no ; or \n characters).

We can't test for all potential hash collisions, so the test set is just the minimum bar to pass, but it must be ensured that an implementation works for any valid data set.

fwbrasil replied:
I'll change the implementation but it'd be nice to have more clear guidelines if you're planning to have another round next year. Most (if not all) of the top contenders have major adaptations to the workload:

1. **Fixed hash table size**: People tune the size to make the workload and tests pass but an execution with more keys could make them enter an infinite loop.
2. **Numeric overflows**: Most implementations use more compact types for things like the sum. It's ok for the current dataset but it could overflow in larger workloads.
3. **Format assumptions**: Several contenders assume the format of the specific workload and would fail if a temperature >= 100 was in the dataset for example.

I'd say finding a good hash function that works well for city names in general in the `long` range seems more reasonable if compared to those assumptions. If latency and performance was a real constraint, I believe that could be a viable tradeoff in a production workload.

gunnarmorling replied:
I hear you in regards to more precisely defined constraints, and yes, it's certainly one of my personal learnings for the next time (no promises yet ;). That said, most of the things you describe in fact are described already:

> Fixed hash table size

There can be 10K stations at most, so a fixed size is fine. You must be able to handle any valid names though, not only those of the example data set.

> Numeric overflows

Do you have an example for that? I suppose a 1B data set with measurements only for one station may overflow an int. Indeed there's no test for that.

> Format assumptions

Temperatures are defined to be between -99.9 and 99.9 (inclusive).

gunnarmorling commented:
Hey @fwbrasil, I am gonna close this one for now as I haven't heard back for a couple of days and I need to clean up the slate a bit. Very welcomed to reopen once all the tests pass and there's no potential for hash collisions. Thanks a lot!
